LXStudio-IDE/src/main/java/titanicsend/pattern/tom/Pulse.java

```diff
@@ -0,0 +1,125 @@
+            LXPoint[] litSection = panelPoints[litIndex];
+            for (LXPoint point : litSection) {
+                colors[point.index] = LXColor.WHITE;
+            }
+
+            for (int i = 0; i < panelPoints.length; i++) {
+                if (i == litIndex) continue;
+                for (LXPoint point : panelPoints[i]) {
+                    colors[point.index] = LXColor.BLACK;
+                }
+            }
+        }
+    }
+
+    private HashMap<String, LXPoint[][]> buildPointMap(HashMap<String, TEPanelModel> panels) {
+        return panels.entrySet().stream()
+                .collect(Collectors.toMap(
+                        e -> e.getKey(),
+                        e -> buildPanelMap(e.getValue()),
+                        (a, b) -> a,
+                        HashMap::new
+                ));
+    }
+
+
+    private LXPoint[][] buildPanelMap(TEPanelModel panel) {
+        ArrayList<ArrayList<LXPoint>> points = new ArrayList<ArrayList<LXPoint>>();
+        Set<LXPoint> unassignedPoints = Arrays.stream(panel.points).collect(Collectors.toSet());
+
+        TEVertex[] currentVertices = {
+                new TEVertex(new LXVector(panel.v0), -1),
+                new TEVertex(new LXVector(panel.v1), -1),
+                new TEVertex(new LXVector(panel.v2), -1)
+        };
+
+        int currentIndex = 0;
+        while (currentVertices[0].distanceTo(panel.centroid) > (10 * PanelStriper.DISTANCE_BETWEEN_PIXELS)) {
+            points.add(new ArrayList<LXPoint>());
+            ArrayList<LXPoint> assignedPoints = new ArrayList<LXPoint>();
+            LXVector[][] edges = {
+                    {currentVertices[0], currentVertices[1]},
+                    {currentVertices[1], currentVertices[2]},
+                    {currentVertices[0], currentVertices[2]},
+            };
+
+            for (LXPoint point : unassignedPoints) {
+                for (LXVector[] edge : edges) {
+                    if (distanceBetPointAndLine(edge[0], edge[1], point) < 5 * PanelStriper.DISTANCE_BETWEEN_PIXELS) {
+                        points.get(currentIndex).add(point);
+                    }
+                }
+            }
+
+            for (LXPoint point : points.get(currentIndex)) {
+                unassignedPoints.remove(point);
+            }
+
+            for (TEVertex vertex : currentVertices) {
+                vertex.nudgeToward(panel.centroid, (float) 0.05);
+            }
+            currentIndex++;
+        }
+
+        return points.stream()
+                .map(l -> l.stream().toArray(LXPoint[]::new))
+                .toArray(LXPoint[][]::new);
+    }
+
+    private static double distanceBetPointAndLine(LXVector line_p1, LXVector line_p2, LXPoint p) {
+        LXVector v0 = new LXVector(p);
+        LXVector v1 = new LXVector(line_p1);
+        LXVector v2 = new LXVector(line_p2);
+        double numerator = new LXVector(v0).sub(v1).cross(new LXVector(v0).sub(v2)).mag();
```

Comment:
v0 is already an LXVector; no need to make it another once twice here

Reply:
Same for lines 117 and 118

Reply:
This is true, but `sub` modifies the vector in place which is why I had to make a copy: https://github.com/heronarts/LX/blob/master/src/main/java/heronarts/lx/transform/LXVector.java#L135.